lib/fd.c

```diff
@@ -0,0 +1,48 @@
+// SPDX-FileCopyrightText: 2024, Skyler Ferrante <sjf5462@rit.edu>
+// SPDX-License-Identifier: BSD-3-Clause
+
+/**
+ * To protect against file descriptor omission attacks, we open the std file
+ * descriptors with /dev/null if they are not already open. Code is based on
+ * fix_fds from sudo.c.
+ */
+
+#include "prototypes.h"
+
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+static void check_fd (int fd);
+
+void
+check_fds (void)
+{
+	/**
+	 * Make sure stdin, stdout, stderr are open
+	 * If they are closed, set them to /dev/null
+	 */
+	check_fd(STDIN_FILENO);
+	check_fd(STDOUT_FILENO);
+	check_fd(STDERR_FILENO);
+}
+
+static void
+check_fd (int fd)
+{
+	int  devnull;
+
+	if (fcntl (fd, F_GETFL, 0) != -1)
+		return;
+
+	devnull = open ("/dev/null", O_RDWR, S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH);
+	if (devnull == -1)
+		abort ();
+
+	if (dup2(devnull, fd) == -1)
+		abort ();
+
+	if (devnull > STDERR_FILENO)
+		close (devnull);
```

Comment:
Hmm, I think you convinced me that this shouldn't happen ever, so let's drop the check?  Or at least abort.  @hallyn , @ikerexxe ?

Reply:
> Static analyzers may also start throwing warnings some day about it.

I think you were right, static analysis tools might give warnings about leaking fds someday if we don't have the check. I can remove it though if you don't want it.

> Or at least abort.

Personally I think closing devnull or just removing the if are probably the best options. Though aborting could work as well.

Reply:
I'd remove the if, that is, the piece of code you presented before I spoke.

Thanks!

Reply:
Okay, I removed the two lines. I also added spaces after all function names (as the rest of the project does).